Add wayland native support with melonds/duckstation method

Miguel-hrvs opened the pull request:
The binary works great with qt6-wayland installed, I just need to fix clang formatting and the appimage needs some qt wayland plugins added to have it working.

SachinVin commented on src/citra_qt/externals/types.h:
```diff
@@ -0,0 +1,39 @@
+
+    melonDS is free software: you can redistribute it and/or modify it under
+    the terms of the GNU General Public License as published by the Free
+    Software Foundation, either version 3 of the License, or (at your option)
```
GPLv3 is not compatible with citra's GPLv2

rtiangha commented:
@Miguel-hrvs I took a look at this and submitted a PR against your branch with some possible improvements/fixes. Could you take a look and try it out?

Miguel-hrvs commented:
Thanks @rtiangha I'll look into it. I ended my exams so now I can get back to these things.

Miguel-hrvs commented:
This could fix wayland only working with xwayland compatibility layer in the appimage: https://github.com/probonopd/linuxdeployqt/issues/406#issuecomment-565463360

rtiangha commented:
High level, the main thing I found was that the new duckstation stuff was linking to `citra` (i.e. the command line app) when it also should have linked to `citra_qt` (i.e. the one with the gui) which is why I moved the code to the `common` folder and linked it to all the executables listed in cmake. But again, I don't have access to a Wayland machine so I can't tell if that actually did anything or if that was the right idea.

rtiangha commented:
Also, I removed the references to `#include "../types.h"` and replaced it with `#include "common/common_types.h"` since it already exists and has everything in it already. And it'll no longer fail the clang-format test that way too. But I think it didn't get merged in which is why the clang-format test is failing.

Miguel-hrvs commented:
> High level, the main thing I found was that the new duckstation stuff was linking to `citra` (i.e. the command line app) when it also should have linked to `citra_qt` (i.e. the one with the gui) which is why I moved the code to the `common` folder and linked it to all the executables listed in cmake. But again, I don't have access to a Wayland machine so I can't tell if that actually did anything or if that was the right idea.

I tested it in wayland in sway wm and only works when you use the binary alone (like before), that's because the appimage doesn't activate wayland qt support if you don't use the -extra-plugins=platforms/ parameter when building it.

Miguel-hrvs commented:
> Also, I removed the references to `#include "../types.h"` and replaced it with `#include "common/common_types.h"` since it already exists and has everything in it already. And it'll no longer fail the clang-format test that way too. But I think it didn't get merged in which is why the clang-format test is failing.

yep, is weird that the clang-format didn't get merged.

rtiangha commented:
> I tested it in wayland in sway wm and only works when you use the binary alone (like before), that's because the appimage doesn't activate wayland qt support if you don't use the -extra-plugins=platforms/ parameter when building it.

I don't know much about AppImages yet, so forgive me if this sounds like a stupid question. When you say you need `-extra-plugins=platforms`, that's a `linuxdeployqt` specific flag, right? From what I can read from `[CMakeModules/BundleTarget.cmake](https://github.com/PabloMK7/citra/blob/master/CMakeModules/BundleTarget.cmake)`, Citra currently uses just `linuxdeploy` with `linuxdeploy-plugin-qt` to make its AppImages. Is there an equivalent way to do the extra-plugins thing with the existing tools? Otherwise, `linuxdeploy` would need to be replaced and the existing options be ported to work with `linuxdeployqt`.

rtiangha commented:
@Miguel-hrvs Just curious, but does [this build artifact](https://github.com/rtiangha/bravely-offline-citra/actions/runs/9664202642/artifacts/1636376838) work under Wayland (note: download will expire in a week)?

Miguel-hrvs commented:
@rtiangha I tested it in sway and still, the only available method in the appimage is xcb. Could be an issue with the package format itself as probonop doesn't like wayland.

rtiangha commented:
You're right. I've been bashing my head trying to figure out why `linuxdeploy-plugin-qt` wouldn't pull in any QT Wayland bits no matter what options I set, and it turns out it doesn't support Wayland at all! Philosophically, as of Apr 24, they're open to supporting it better welcoming any PRs that'll help, but that doesn't help things right now or for the foreseeable future

I suppose I could try to figure out how to get `linuxdeployqt` to work for the GUI version instead, but before I do, what library files specifically need to be copied into the App image? I'd like to try manually copying them in to see if that'll work and as a sanity check, but again, I don't have access to a Wayland machine to test with.

rtiangha commented:
@Miguel-hrvs OK, looks like [this one](https://github.com/rtiangha/bravely-offline-citra/actions/runs/9929880649/artifacts/1699639796) pulled in some QT Wayland bits, so please try it; I think it might work (artifact expires in one week).

Miguel-hrvs commented:
@rtiangha. Now I'm in Athens. I'll come back thursday at night. So I'll test it the next day. Cheers :)

Miguel-hrvs commented:
Please, send this pr of yours to my wayland branch: https://github.com/rtiangha/bravely-offline-citra/commit/8ee3a6bc7b3df35b67b828bae8030ff6c899e3ba

rtiangha commented:
Done. Although I wonder if adding those options to the Linux ci script was all that was needed to begin with. Regardless, nice to hear that it now works.

Miguel-hrvs commented:
Seems that those ci changes where what we needed to get it working in the appimages. Ready for review, thanks again :)

rtiangha commented:
So the duckstation stuff wasn't needed in the first place; it was just the ci change?

Miguel-hrvs commented:
For appimage to add wayland support yes. But the egl, glad and khr additions where needed too.

rtiangha commented:
So shouldn't the PR be changed to remove the stuff that really isn't necessary?

Miguel-hrvs commented:
I'll take a second look into what's needed to be sure

Miguel-hrvs commented:
@rtiangha omg it really was just needed the two line ci change. I'll open a different pr with just that change. A guy told me that something like this was what we needed but didn't now how to do it. Now we have it, feels so easy now XD.

rtiangha commented:
Well, it's good to know for certain then. It may not have worked in the past as linuxdeploy-plugin-qt's Wayland support back then may not have pulled in everything else that was needed to work.
